### ISSUE TITLE
Randomly, Application Inspector analysis never ends.

### ISSUE THREAD

**Issue by ismadirolas** [microsoft/ApplicationInspector#436]
**Describe the bug**
We have experienced sometimes, apparently random, that applications inspector analysis stucks the machine (we think maybe it is a memory problem) and never ends. We always pass to the cli the argument --process-timeout, but in this cases the execution never ends. What do you think it's going on? Thanks for the help.

**Expected behavior**
The execution of an analysis always respects the process-timeout param time.

**Additional context**
It happens when execute the application inspector inside a github or a bamboo workflow.

**Comment by gfs** [comment]
Hey there,

Sorry you're having trouble. I have some questions that might help me figure the issue. 

- What is the full command line argument you're using?
- Approximately how many files/what language is what you're scanning? If it's open source and you can link me to it that's even better.
- what is the console output when it stalls?

**Comment by ismadirolas** [comment]
Hi @gfs

- The full command line is like `./ApplicationInspector.CLI analyze -s . -g **/rulesFolder/**,**/.github/actions/**,**/node_modules/**,**/.git/**,**/target/**,**/package-lock.json,*.mock*,*.jpg,*.png,*.gif,*.woff,*.woff2,*.svg,*.ttf -f json --processing-timeout 600000 -o result.json -A -c high,medium,low -u -r rulesFolder -i -M 20`
- The analyzed projects are not open source, sorry. For example, we get the next results with a project that sometimes has the error:
`"totalFiles": 19002,
  "filesTimedOut": 0,
  "filesAnalyzed": 2322,
  "filesSkipped": 16680,
  "filesAffected": 1627,
  "filesErrored": 0,
  "totalMatchesCount": 116,
  "uniqueMatchesCount": 9`

`"fileExtensions": [
      "",
      "afm",
      "aliases",
      "api_description",
      "BridgeExtension",
      "cat",
      "cpp",
      "css",
      "DatatypeFactory",
      "def",
      "DocumentBuilderFactory",
      "DomExtension",
      "DOMImplementationSourceList",
      "driver",
      "Driver",
      "dtd",
      "ecore",
      "escaped",
      "exsd",
      "genmodel",
      "gitattributes",
      "gitignore",
      "gperf",
      "htm",
      "html",
      "inf",
      "ini",
      "InterpreterFactory",
      "java",
      "jj",
      "jjt",
      "js",
      "json",
      "jsp",
      "list",
      "Logger",
      "mappings",
      "mdl",
      "MF",
      "mod",
      "options",
      "permissions",
      "policy",
      "profile",
      "properties",
      "ps",
      "rels",
      "res",
      "rng",
      "rptdesign",
      "SAXParserFactory",
      "SchemaFactory",
      "SF",
      "snyk",
      "SquiggleInputHandler",
      "tld",
      "txt",
      "xdt",
      "xmi",
      "xml",
      "xsd",
      "xsd2ecore"
    ],
    "cloudTargets": [],
    "CPUTargets": [],
    "languages": {
      "build.xml": 1,
      "cpp": 1,
      "html": 263,
      "java": 751,
      "javascript": 10,
      "json": 2,
      "pom.xml": 3,
      "Unknown": 1291
    },`

- There is no output when it happens, so we don't get even the "Analyze command running" message.

**Comment by gfs** [comment]
Thanks for the extra info. 

I'll take a look today.

**Comment by gfs** [comment]
I do not have the custom rules you have so I cannot validate if it's a problem with those in particular.

I have tested the overall timeout with files that I have and it seems to work fine locally.  You seemed to indicate that this is just a CI issue so it looks like I need to test in a github workflow too.

Once issue I have noticed is that currently if you are using the progress bar (default) the message that time out has occurred wont be visible.

Have you tried with `--no-show-progress`?  It might be the case that the progress bar display does not work well in a CI environment - it is designed for interactive use.

I also have a PR which will improve the output you get if you hit a timeout: #437.

**Comment by gfs** [comment]
Hi there. I have some additional follow up questions:

* Do you only see this with your custom rules or with default application inspector rules too?

* Does the timeout work if you set it very low? For example a 1 ms timeout should instantly stop.

* Do you only see this in CI environments or also when running locally?

**Comment by ismadirolas** [comment]
Hi, we are going to try the new version with the improved logs and I will reply to you. Thanks.

**Comment by gfs** [comment]
I am working on a change that I'm planning to merge this week that will revamp the logging further if you are still having issues.

**Comment by gfs** [comment]
Hi @ismadirolas are you still having trouble? 

The rewrite ended up getting larger than I had originally planned and is in #441.

**Comment by ismadirolas** [comment]
Hi @gfs , 
Yes, we still have trouble sometimes, we have done a walkaround with an external timeout, but will be interesting to check the new logs when you merge the PR.

**Comment by gfs** [comment]
> Hi @gfs , 
> 
> Yes, we still have trouble sometimes, we have done a walkaround with an external timeout, but will be interesting to check the new logs when you merge the PR.

Thanks for letting me know.

I have some time to do some more on this this week so hopefully will have some movement on the pr soon.

**Comment by gfs** [comment]
1.5-beta is available now if you want to check if it helps with reliability 

You should also now be able to set verbose logging to a file and run single threaded to identify the specific file(s) that are stalling.

**Comment by ismadirolas** [comment]
Hi,
we have notice that if we exclude .zip and .jar files, it works better.
Thanks for the help.

**Comment by gfs** [comment]
> Hi,
> 
> we have notice that if we exclude .zip and .jar files, it works better.
> 
> Thanks for the help.

Thanks. That is very useful information to have.

Application Inspector will try to scan the files inside of most archives types if they are not excluded (anything supported by https://github.com/microsoft/RecursiveExtractor). A jar is actually just a zip with a different extension, so that pretty conclusively points to it being archive related.

If those archives are large the set being scanned may actually be much larger than you are anticipating, and the long run time is correct. However, that does give me a hint about somewhere there might be an edge case where the timer doesn't get hit, and recursive extractor has its own timeout mechanism that I don't believe AI is currently using that I could also potentially configure.

I'll also add an option to 1.5 for skipping the archive crawling behavior entirely without having to exhaustively list archive extensions in the path exclusions.

**Comment by gfs** [comment]
I think I've identified the issue. 


Application Inspector has three steps:
Enumerate the things to scan
Scan the things
Prepare the report

If you use the progress bar the enumeration step happens first, and then processing happens. If you disable the progress bar the processing step enumerates as it goes. 

The timeout is correctly, but perhaps not clearly, labeled processing timeout and it only applies to the processing step.

If you encounter a hang in the enumerating step or the preparing report step the timeout does not cover it. 

You note that this happens with zip and jar files, which potentially implicates a hang in the enumerating step. However, I actually was able to reproduce a hang in the report preparation step due to an out of memory error if the report is very large and you are using the HTML report.

I'm working on Application inspector this week and I'll be trying to address this issue in 1.5 now that I've identified potential additions

1. Total overall timeout
2. Enumeration timeout
3. Report preparation timeout
4. Catch errors in html report preparation, determine fallback behavior if html report is too large to construct.
